patches/tModLoader/Terraria/ModLoader/Core/ModCompile.cs

```diff
@@ -439,29 +475,38 @@ private string DllRefPath(BuildingMod mod, string dllName)
 		}
 
 		/// <summary>
-		/// Compile a dll for the mod based on required includes.
+		/// Load a .csproj from csprojPath, using the <paramref name="preprocessorSymbols"/> to determine the optimization
+		/// level, and allowing unsafe code depending on <paramref name="allowUnsafe"/>
 		/// </summary>
-		private static Diagnostic[] RoslynCompile(string name, List<string> references, string[] files, string[] preprocessorSymbols, bool allowUnsafe, out byte[] code, out byte[] pdb)
-		{
-			var options = new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary,
-				assemblyIdentityComparer: DesktopAssemblyIdentityComparer.Default,
-				optimizationLevel: preprocessorSymbols.Contains("DEBUG") ? OptimizationLevel.Debug : OptimizationLevel.Release,
-				allowUnsafe: allowUnsafe);
+		private static Project LoadProject(string csprojPath, string[] preprocessorSymbols, bool allowUnsafe) {
+			using var workspace = MSBuildWorkspace.Create();
+			var project = workspace.OpenProjectAsync(csprojPath).Result;
 
-			var parseOptions = new CSharpParseOptions(LanguageVersion.Preview, preprocessorSymbols: preprocessorSymbols);
+			if (project.CompilationOptions is not null) {
+				var newOptions = project.CompilationOptions
+					.WithAssemblyIdentityComparer(DesktopAssemblyIdentityComparer.Default)
+					.WithOptimizationLevel(preprocessorSymbols.Contains("DEBUG") ? OptimizationLevel.Debug : OptimizationLevel.Release);
 
-			var emitOptions = new EmitOptions(debugInformationFormat: DebugInformationFormat.PortablePdb);
+				if (project.CompilationOptions is CSharpCompilationOptions csharpComp)
+					newOptions = csharpComp.WithAllowUnsafe(allowUnsafe);

```

Comment:
Is overwriting the above assigned `newOptions` intentional?

Reply:
Yes, it's intentional. The `newOptions` is of type `CompilationOptions`, which doesn't have an `allowUnsafe` field, so in case `newOptions` is of type `CSharpCompilationOptions`, we cast it and set the `allowUnsafe` field to whatever was passed to TML. I could make it so it always assumes it is going to compile a C# project, but it would throw an error if it tried to compile a VB.NET or F# project, and I'd prefer to leave it like this just in case someone wanted to write a mod using another .NET language.

Reply:
The thing is, calling `CompilationOptions.WithX` doesn't modify the `CompilationOptions`.
Meaning `newOptions = csharpComp.WithAllowUnsafe(allowUnsafe);` overwrites/deletes the options you set for `newOptions`.

What the current code does is:
1. Create `newOptions` with `AssemblyIdentityComparer` and `OptimizationLevel`
2. If `project.CompilationOptions is CSharpCompilationOptions`, completely ignore the above actions and only set `AllowUnsafe`

Reply:
The fix for that would be to replace
`if (project.CompilationOptions is CSharpCompilationOptions csharpComp)`
with
`if (newOptions is CSharpCompilationOptions csharpComp)`

Reply:
ohhhh, I didn't see that, thanks for letting me know

Reply:
Should be fixed now

Reply:
Cool. Close this thread please